src/main/java/net/coderbot/iris/pipeline/TransformPatcher.java

```diff
@@ -0,0 +1,512 @@
+										.collect(Collectors.toList()));
+			}
+
+			@Override
+			protected ActivatableLifecycleUser<Parameters> getInjector() {
+				return RunPhase.withInjectExternalDeclarations(injectionLocation(),
+						"uniform vec2 u_TextureScale;",
+						"attribute vec2 a_TexCoord;");
+			}
+		}
+				.wrapTarget("gl_MultiTexCoord0")
+				.parsedReplacement("vec4(a_TexCoord * u_TextureScale, 0.0, 1.0)")
+				.injectionLocation(InjectionPoint.BEFORE_DECLARATIONS);
+
+		LifecycleUser<Parameters> wrapColor = new WrapIdentifier<Parameters>()
+				.wrapTarget("gl_Color")
+				.detectionResult("a_Color")
+				.injectionLocation(InjectionPoint.BEFORE_DECLARATIONS)
+				.injectionExternalDeclaration("attribute vec4 a_Color;");
+
+		LifecycleUser<Parameters> wrapNormal = new WrapIdentifier<Parameters>()
+				.wrapTarget("gl_Normal")
+				.detectionResult("a_Normal")
+				.injectionLocation(InjectionPoint.BEFORE_DECLARATIONS)
+				.injectionExternalDeclaration("attribute vec3 a_Normal;");
+
+		LifecycleUser<Parameters> wrapModelViewMatrix = new WrapIdentifier<Parameters>()
+				.wrapTarget("gl_ModelViewMatrix")
+				.detectionResult("u_ModelViewMatrix")
+				.injectionLocation(InjectionPoint.BEFORE_DECLARATIONS)
+				.injectionExternalDeclaration("uniform mat4 u_ModelViewMatrix;");
+
+		LifecycleUser<Parameters> wrapModelViewProjectionMatrix = new WrapIdentifier<Parameters>()
+				.wrapTarget("gl_ModelViewProjectionMatrix")
+				.detectionResult("u_ModelViewProjectionMatrix")
+				.injectionLocation(InjectionPoint.BEFORE_DECLARATIONS)
+				.injectionExternalDeclaration("uniform mat4 u_ModelViewProjectionMatrix;");
+
+		LifecycleUser<Parameters> wrapNormalMatrix = new WrapIdentifier<Parameters>()
+				.wrapTarget("gl_NormalMatrix")
+				.detectionResult("u_NormalMatrix")
+				.parsedReplacement("mat3(u_NormalMatrix)")
+				.injectionLocation(InjectionPoint.BEFORE_DECLARATIONS)
+				.injectionExternalDeclaration("uniform mat4 u_NormalMatrix;");
+
+		LifecycleUser<Parameters> replaceTextureMatrix0 = new Transformation<Parameters>() {
+			{
+				addEndDependent(new WalkPhase<Parameters>() {
+					ParseTreePattern textureMatrixPattern;
+
+					@Override
+					public void init() {
+						textureMatrixPattern = compilePattern("gl_TextureMatrix[0]", GLSLParser.RULE_expression);
+					}
+
+					@Override
+					public void enterArrayAccessExpression(ArrayAccessExpressionContext ctx) {
+						ParseTreeMatch match = textureMatrixPattern.match(ctx);
+						if (match.succeeded()) {
+							replaceNode(ctx, "mat4(1.0)", GLSLParser::expression);
+						}
+					}
+				});
+			}
+		};
+
+		/**
+		 * Implements BuiltinUniformReplacementTransformer and does a little more. Note
+		 * that the main walk phase uses the fact that the order of invocation is
+		 * enterMemberAccessExpression, enterMultiplicativeExpression,
+		 * enterArrayAccessExpression in the targeted expression.
+		 */
+		LifecycleUser<Parameters> replaceBuiltinUniforms = new Transformation<Parameters>() {
```

Comment:
"Builtin uniforms replacement" is actually a misnomer that probably shouldn't be carried over

Reply:
What would an appropriate name be for this transformation?

Reply:
replaceLightmapForSodium or similar could work